### PR TITLE
Sync stable-cadence branch with master

### PR DESCRIPTION
**Pull request by SupunS**
## Description

$subject

______

- [ ] Targeted PR against `master` branch
- [x] Linked to Github issue with discussion and accepted design OR link to spec that describes this work
- [x] Code follows the [standards mentioned here](https://github.com/onflow/cadence/blob/master/CONTRIBUTING.md#styleguides)
- [x] Updated relevant documentation 
- [x] Re-reviewed `Files changed` in the Github PR explorer
- [x] Added appropriate labels

### CODE REVIEW

File: runtime/runtime_test.go [hunk 1]
```diff
@@ -7979,8 +7994,141 @@ func TestRuntimeInvalidatedResourceUse(t *testing.T) {
...
+		},
+	)
+	RequireError(t, err)
+
+	assertRuntimeErrorIsUserError(t, err)
+
+	var parserError parser.Error
+	require.ErrorAs(t, err, &parserError)
 
+	assert.IsType(t, &parser.CustomDestructorError{}, parserError.Errors[0])
```
Comment (SupunS) [inline runtime/runtime_test.go:8131]: This and the below ones are no longer valid, since destructors are no longer supported and those edge cases were only possible to be reached in the destructor. So updated the test to assert the new static error (`CustomDestructorError`).

We could technically move these tests to checker tests, but in future, if/when custom destructors are allowed, these would become valid tests again, so keeping it here would make life easier + won't miss out on adding these back again.

File: runtime/sema/check_interface_declaration.go [hunk 2]
```diff
@@ -33,6 +33,12 @@ import (
 // through `declareInterfaceMembers`.
 func (checker *Checker) VisitInterfaceDeclaration(declaration *ast.InterfaceDeclaration) (_ struct{}) {
 
+	wasInInterface := checker.inInterface
+	checker.inInterface = true
+	defer func() {
+		checker.inInterface = wasInInterface
+	}()
```
Comment (SupunS) [inline runtime/sema/check_interface_declaration.go:40]: This was removed in https://github.com/dapperlabs/cadence-internal/pull/179, but had to add this back again, since entitlements were relying on it.
We should double-check if an issue similar to the original one also exists with entitlements. cc: @dsainati1

File: runtime/tests/interpreter/resources_test.go [hunk 3]
```diff
@@ -2137,158 +2137,6 @@ func TestInterpretResourceDestroyedInPreCondition(t *testing.T) {
 	require.ErrorAs(t, err, &interpreter.InvalidatedResourceError{})
 }
 
-func TestInterpretInvalidReentrantResourceDestruction(t *testing.T) {
```
Comment (SupunS) [inline runtime/tests/interpreter/resources_test.go]: Removed this test since this is no longer valid, after destructor removal

File: runtime/tests/interpreter/resources_test.go [hunk 4]
```diff
@@ -2658,205 +2506,6 @@ func TestInterpretDefaultDestroyEventArgumentScoping(t *testing.T) {
 	require.Equal(t, interpreter.NewIntValueFromInt64(nil, 1), events[0].GetField(inter, interpreter.EmptyLocationRange, "x"))
 }
 
-func TestInterpretInnerResourceMove(t *testing.T) {
```
Comment (SupunS) [inline runtime/tests/interpreter/resources_test.go]: Same here: Removed this test since this is no longer valid, after destructor removal

File: runtime/tests/interpreter/resources_test.go [hunk 5]
```diff
@@ -2658,205 +2506,6 @@ func TestInterpretDefaultDestroyEventArgumentScoping(t *testing.T) {
...
-            }`,
-		)
-
-		_, err := inter.Invoke("main")
-		RequireError(t, err)
-		require.ErrorAs(t, err, &interpreter.UseBeforeInitializationError{})
-	})
-}
-
-func TestInterpretSetMemberResourceLoss(t *testing.T) {
```
Comment (SupunS) [inline runtime/tests/interpreter/resources_test.go]: Removed this test since this is no longer valid, after destructor removal

File: runtime/tests/interpreter/resources_test.go [hunk 6]
```diff
@@ -2658,205 +2506,6 @@ func TestInterpretDefaultDestroyEventArgumentScoping(t *testing.T) {
...
-       destroy rl
-    }`,
-	)
-
-	_, err := inter.Invoke("main")
-	RequireError(t, err)
-	require.ErrorAs(t, err, &interpreter.DestroyedResourceError{})
-}
-
-func TestInterpretValueTransferResourceLoss(t *testing.T) {
```
Comment (SupunS) [inline runtime/tests/interpreter/resources_test.go]: Moved to runtime tests, since the custom destructors are not allowed, destruction of a resource can only be asserted from the emitted event.

File: runtime/tests/interpreter/resources_test.go [hunk 7]
```diff
@@ -2934,43 +2583,14 @@ func TestInterpretVariableDeclarationEvaluationOrder(t *testing.T) {
 	)
 }
 
-func TestInterpretIfLetElseBranchConfusion(t *testing.T) {
```
Comment (SupunS) [inline runtime/tests/interpreter/resources_test.go]: This gives a checker error now, so moved to checker tests

File: runtime/tests/interpreter/resources_test.go [hunk 8]
```diff
@@ -3039,41 +2657,3 @@ func TestInterpretMovedResourceInSecondValue(t *testing.T) {
 	assert.Equal(t, 15, errorStartPos.Line)
 	assert.Equal(t, 53, errorStartPos.Column)
 }
-
-func TestInterpretOptionalBindingElseBranch(t *testing.T) {
```
Comment (SupunS) [inline runtime/tests/interpreter/resources_test.go]: This gives a checker error now, so moved to checker tests

File: runtime/tests/checker/resources_test.go [hunk 9]
```diff
@@ -9526,61 +9529,6 @@ func TestCheckInvalidationInCondition(t *testing.T) {
 				))
 				require.NoError(t, err)
 			})
-
-			t.Run("in nested type requirement, definite invalidation", func(t *testing.T) {
```
Comment (SupunS) [inline runtime/tests/checker/resources_test.go]: No longer valid after the type requirement removal

**Review by turbolent — APPROVED** [review]
Perfect! Thank you for adding the comments with the reasoning for (re)moving the test cases 🙏

**Review by dsainati1 — APPROVED** [review]
